srk/src/almostCommuting.ml

```diff
@@ -0,0 +1,146 @@
+    let mSS = intersect_rowspaces 
+                (Array.map (fun m -> QQMatrix.mul m mS) sims)
+                dims
+    in
+    if rowspace_equal mS mSS then
+      mS, matr
+    else
+      fix mSS
+  in
+  fix mS
+
+let iter_all     = Array.map (fun (_, m) -> m)
+let iter_reset   = BatArray.filter_map (fun (k, m) -> if k == Reset then Some m else None)
+let iter_commute = BatArray.filter_map (fun (k, m) -> if k == Commute then Some m else None)
+
+module PhasedSegment = struct
+
+  type t = phased_segment
+
+  (* TODO: find proper representation *)
+  let show p = QQMatrix.show p.sim1
+
+  let equal p q =
+    QQMatrix.equal p.sim1 q.sim1 &&
+    QQMatrix.equal p.sim2 q.sim2 &&
+    BatArray.for_all2
+      QQMatrix.equal
+      p.phase1
+      q.phase1 &&
+    BatArray.for_all2
+      (fun (k1, m1) (k2, m2) -> k1 == k2 && QQMatrix.equal m1 m2)
+      p.phase2
+      q.phase2
+
+  let make pairs =
+    if Array.length pairs == 0 then
+      raise (Invalid_argument "Array of matrices should not be empty")
+    else
+      let _, mA = Array.get pairs 0 in
+      let dims = SrkUtil.Int.Set.elements (QQMatrix.column_set mA) in
+      let mS, phase1 = commuting_segment (iter_all pairs) dims in
+      let mT, _ = commuting_segment (iter_commute pairs) dims in
+      let rec fix mT =
+        let ls_maxlds = Array.map
+          (fun (k, m) ->
+            if k == Reset then
+              max_lds ~zero_rows:true mS (QQMatrix.mul mT m)
+            else if k == Commute then
+              max_lds mT (QQMatrix.mul mT m)
+            else
+              mT, m)
+          pairs
+        in
+        let mTT = intersect_rowspaces
+                    (Array.map (fun (m, _) -> QQMatrix.mul m mT) ls_maxlds)
+                    dims
+        in
+        if rowspace_equal mT mTT then
+          let phase2 = Array.mapi
+                         (fun i (_, m) -> let k, _ = Array.get pairs i in (k, m))
+                         ls_maxlds
+          in
+          { sim1 = mS;
+            sim2 = mTT;
+            phase1 = phase1;
+            phase2 = phase2 }
+        else
+          fix mTT
+      in
+      fix mT
+
+end
+
+module PhasedSegmentation = struct
+
+  type t = phased_segment list
+
+  let make_naive matrices = 
+    let len = Array.length matrices in
+    let products = BatList.n_cartesian_product (BatList.make len [Commute; Reset]) in
+    let partitions = BatList.map 
+                       (fun p -> Array.map2 (fun x y -> x, y) (Array.of_list p) matrices) 
+                       products
+    in
+    BatList.map PhasedSegment.make partitions
+
+  let almost_commuting_space segmentation =
+    List.fold_left
+      (fun vU s -> VS.sum vU (VS.of_matrix s.sim2))
+      VS.empty
+      segmentation
+
+  let almost_commutes segmentation =
+    let vU = almost_commuting_space segmentation in
+    let dim = VS.dimension vU in
+    VS.equal vU (VS.standard_basis dim)
+
+  let rec best_almost_commuting matrices =
```

Comment:
1. Does this need to be recursive?
2. I think this should return an LTS + simulation rather than a phased segmentation of that LTS
3. divide_right (mC * m) mC will give you a the correct (square) matrix:
(divide_right (mC * m) mC) * mC = mC * m

Reply:
I changed it according to your suggestions.